File: backend/app/routes/comments.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import db, Comment, Recipe, User

comments_bp = Blueprint('comments', __name__)

@comments_bp.route('/recipe/<int:recipe_id>', methods=['GET'])
def get_recipe_comments(recipe_id):
    """获取指定菜谱的所有评论"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 10, type=int)

    # 验证菜谱是否存在
    recipe = Recipe.query.get_or_404(recipe_id)

    comments = Comment.query.filter_by(recipe_id=recipe_id)\
        .order_by(Comment.created_at.desc())\
        .paginate(page=page, per_page=per_page, error_out=False)

    return jsonify({
        'comments': [comment.to_dict() for comment in comments.items],
        'total': comments.total,
        'pages': comments.pages,
        'current_page': page
    })

@comments_bp.route('/', methods=['POST'])
@jwt_required()
def create_comment():
    """创建新评论"""
    data = request.get_json()
    user_id = get_jwt_identity()

    # 验证必填字段
    if not data.get('content'):
        return jsonify({'message': '评论内容不能为空'}), 400

    if not data.get('recipe_id'):
        return jsonify({'message': '菜谱ID不能为空'}), 400

    # 验证菜谱是否存在
    recipe = Recipe.query.get_or_404(data['recipe_id'])

    # 验证评分范围
    rating = data.get('rating', 5)
    if not isinstance(rating, int) or rating < 1 or rating > 5:
        return jsonify({'message': '评分必须在1-5之间'}), 400

    comment = Comment(
        content=data['content'],
        rating=rating,
        user_id=user_id,
        recipe_id=data['recipe_id']
    )

    db.session.add(comment)
    db.session.commit()

    # 返回评论详情，包含用户信息
    comment_data = comment.to_dict()
    comment_data['user'] = User.query.get(user_id).to_dict()

    return jsonify(comment_data), 201

@comments_bp.route('/<int:comment_id>', methods=['PUT'])
@jwt_required()
def update_comment(comment_id):
    """更新评论（只有评论作者可以更新）"""
    user_id = get_jwt_identity()
    user = User.query.get(user_id)
    comment = Comment.query.get_or_404(comment_id)

    # 检查权限（只有评论作者可以编辑）
    if comment.user_id != user_id:
        return jsonify({'message': '无权限编辑此评论，只有评论作者可以编辑'}), 403

    data = request.get_json()

    # 更新字段
    if 'content' in data:
        comment.content = data['content']

    if 'rating' in data:
        rating = data['rating']
        if not isinstance(rating, int) or rating < 1 or rating > 5:
            return jsonify({'message': '评分必须在1-5之间'}), 400
        comment.rating = rating

    db.session.commit()
    return jsonify(comment.to_dict())

@comments_bp.route('/<int:comment_id>', methods=['DELETE'])
@jwt_required()
def delete_comment(comment_id):
    """删除评论（评论作者、菜谱创建者或管理员可以删除）"""
    user_id = get_jwt_identity()
    user = User.query.get(user_id)
    comment = Comment.query.get_or_404(comment_id)

    # 检查权限（评论作者、菜谱创建者或管理员可以删除）
    if not user.can_manage_comment(comment):
        return jsonify({'message': '无权限删除此评论，只有评论作者、菜谱创建者或管理员可以删除'}), 403

    db.session.delete(comment)
    db.session.commit()
    return jsonify({'message': '评论删除成功'})

@comments_bp.route('/user/<int:user_id>', methods=['GET'])
def get_user_comments(user_id):
    """获取指定用户的所有评论"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 10, type=int)

    # 验证用户是否存在
    user = User.query.get_or_404(user_id)

    comments = Comment.query.filter_by(user_id=user_id)\
        .order_by(Comment.created_at.desc())\
        .paginate(page=page, per_page=per_page, error_out=False)

    # 添加菜谱信息到评论
    comments_data = []
    for comment in comments.items:
        comment_dict = comment.to_dict()
        comment_dict['recipe'] = Recipe.query.get(comment.recipe_id).to_dict()
        comments_data.append(comment_dict)

    return jsonify({
        'comments': comments_data,
        'total': comments.total,
        'pages': comments.pages,
        'current_page': page
    })

@comments_bp.route('/recipe/<int:recipe_id>/stats', methods=['GET'])
def get_recipe_comment_stats(recipe_id):
    """获取菜谱的评论统计信息"""
    # 验证菜谱是否存在
    recipe = Recipe.query.get_or_404(recipe_id)

    # 获取评论数量
    total_comments = Comment.query.filter_by(recipe_id=recipe_id).count()

    if total_comments == 0:
        return jsonify({
            'total_comments': 0,
            'average_rating': 0,
            'rating_distribution': {
                1: 0, 2: 0, 3: 0, 4: 0, 5: 0
            }
        })

    # 计算平均评分
    avg_rating = db.session.query(db.func.avg(Comment.rating))\
        .filter_by(recipe_id=recipe_id).scalar() or 0

    # 获取评分分布
    rating_distribution = {}
    for rating in range(1, 6):
        count = Comment.query.filter_by(recipe_id=recipe_id, rating=rating).count()
        rating_distribution[rating] = count

    return jsonify({
        'total_comments': total_comments,
        'average_rating': round(float(avg_rating), 2),
        'rating_distribution': rating_distribution
    })

File: frontend/src/api/comments.ts
import api from './index'
import { type User } from './auth'

// 评论类型定义
export interface Comment {
  id: number
  content: string
  user_id: number
  recipe_id: number
  parent_id?: number
  created_at: string
  updated_at?: string
  user?: User
  replies?: Comment[]
}

// 评论列表响应类型
export interface CommentsResponse {
  comments: Comment[]
  total: number
  pages: number
  current_page: number
}

// 评论统计类型
export interface CommentStats {
  comment_count: number
  rating_count: number
  average_score: number
  rating_distribution: {
    1: number
    2: number
    3: number
    4: number
    5: number
  }
}

// 创建评论请求数据
export interface CreateCommentRequest {
  content: string
  rating?: number
}

// 更新评论请求数据
export interface UpdateCommentRequest {
  content?: string
}

// 评论相关API
export const commentApi = {
  // 获取菜谱评论列表
  getRecipeComments(recipeId: number, query?: { page?: number; per_page?: number }) {
    const params = new URLSearchParams()
    if (query?.page) params.append('page', query.page.toString())
    if (query?.per_page) params.append('per_page', query.per_page.toString())

    const queryString = params.toString()
    const url = queryString ? `/recipes/${recipeId}/comments?${queryString}` : `/recipes/${recipeId}/comments`

    return api.get<CommentsResponse>(url)
  },

  // 创建评论
  createComment(recipeId: number, data: CreateCommentRequest) {
    return api.post<Comment>(`/recipes/${recipeId}/comments`, data)
  },

  // 更新评论
  updateComment(commentId: number, data: UpdateCommentRequest) {
    return api.put<Comment>(`/comments/${commentId}`, data)
  },

  // 删除评论
  deleteComment(commentId: number) {
    return api.delete<{ message: string }>(`/comments/${commentId}`)
  },

  // 获取菜谱评论统计
  getRecipeCommentStats(recipeId: number) {
    return api.get<CommentStats>(`/recipes/${recipeId}/comments/stats`)
  }
}

// 评分相关工具函数
export const ratingUtils = {
  // 获取评分星星数组
  getRatingStars(rating: number): boolean[] {
    const stars = []
    for (let i = 1; i <= 5; i++) {
      stars.push(i <= rating)
    }
    return stars
  },

  // 获取评分文本
  getRatingText(rating: number): string {
    const texts = ['', '很差', '较差', '一般', '推荐', '强烈推荐']
    return texts[rating] || '未评分'
  },

  // 获取评分颜色
  getRatingColor(rating: number): string {
    if (rating >= 4.5) return '#67c23a'
    if (rating >= 3.5) return '#e6a23c'
    if (rating >= 2.5) return '#f56c6c'
    return '#909399'
  }
}

// 格式化相对时间
export const formatRelativeTime = (dateString: string): string => {
  const date = new Date(dateString)
  const now = new Date()
  const diffInSeconds = Math.floor((now.getTime() - date.getTime()) / 1000)

  if (diffInSeconds < 60) return '刚刚'
  if (diffInSeconds < 3600) return `${Math.floor(diffInSeconds / 60)}分钟前`
  if (diffInSeconds < 86400) return `${Math.floor(diffInSeconds / 3600)}小时前`
  if (diffInSeconds < 604800) return `${Math.floor(diffInSeconds / 86400)}天前`

  return date.toLocaleDateString('zh-CN', {
    year: 'numeric',
    month: 'long',
    day: 'numeric'
  })
}